enhancements/update/change-management-and-maintenance-schedules.md

```diff
@@ -0,0 +1,1334 @@
+1. Providing strict promises around the exact timing of upgrade processes. Maintenance schedules will be honored to a 
+   reasonable extent (e.g. upgrade actions will only be initiated during a window), but long-running operations may 
+   exceed the configured end of a maintenance schedule.
+1. Implementing logic to defend against impractical maintenance schedules (e.g. if a customer configures a 1-second 
+   maintenance schedule every year). Service Delivery may want to implement such logic to ensure upgrade progress can 
+   be made.
+1. Automatically initiating updates to `ClusterVersion`. This will still occur through external actors/orchestration. 
+   Maintenance schedules simply give the assurance that changes to `ClusterVersion` will not result in material changes 
+   until permitted by the defined maintenance schedules.
+
+## Proposal
+
+### Change Management Overview
+Establish a new namespaced Custom Resource Definition (CRD) called `ChangeManagementPolicy` which allows cluster lifecycle
+administrators to capture their requirements for when resource(s) associated with the policy can initiate material 
+changes on the cluster.
+
+Add a `changeManagement` stanza to several existing resources in the OpenShift ecosystem which can reference
+the new `ChangeManagementPolicy` resource to restrict when and how their associated controllers can initiate
+material changes to a cluster.
+- HCP's `HostedCluster`. Honored by HyperShift Operator and supported by underlying CAPI providers.
+- HCP's `NodePool`. Honored by HyperShift Operator and supported by underlying CAPI providers.
+- Standalone's `ClusterVersion`. Honored by Cluster Version Operator.
+- Standalone's `MachineConfigPool`. Honored by Machine Config Operator.
+
+Changes that are not allowed to be initiated due to a change management policy will be 
+called "pending". Controllers responsible for initiating pending changes will await a permissive window 
+according to each resource's relevant `changeManagement` configuration.
+
+In additional to "policies", different resource kinds may offer additional knobs in their `changeManagement`
+stanzas to provide cluster lifecycle administrators additional control over the exact nature of 
+the changes desired by the resource's associated controller. For example, in `MachineConfigPool`'s 
+`changeManagement` stanza, a cluster lifecycle administrator will be able to (optionally) specify the 
+exact rendered configuration the controller should be working towards (during the next permissive window) vs
+the traditional OpenShift model where the "latest" rendered configuration is always the destination. 
+
+### ChangeManagementPolicy Resource
+
+```yaml
+kind: ChangeManagementPolicy
+metdata:
+  # ChangeManagementPolicies are namespaced resources. They will normally reside in the 
+  # namespace associated with the controller initiating material changes. 

```

Comment:
The actor would be namespaced for HCP. So unless we want to have a namespaced (`HostedChangeManagementPolicy`?) and non-namespaced version of the CRD, namespaced was my compromise.

Reply:
A separate, namespace scoped version could be achievable, it may require some tooling improvements to make it easier to manage, but, nothing is impossible right?

Reply:
> you will need a different set of fields for the HCP use vs OCP use
Anything more on how the fields would be different? This is just the policy definition right?

Reply:
So typically this would be to do with references, if the resource references a secret, or a configmap, or some other object, typically a namespaced resource does not include a namespace field in the reference, as the referenced object must always be in the same namespace. For a cluster scoped, you have to specify the namespace.

In this API, it's different though. We don't have any namespaced references inside the API itself, so it would be the same spec and status needed, vis-a-vis some minor documentation differences.

The issue though is on the binding side. At the moment, the expectation is that the binding is by referencing the `ChangeManagementPolicy` from objects like `MachineConfig`, which is cluster scoped. So what do we do with that reference?

It would be the one that needed to be different, at which point, we are creating way too much complexity IMO.

How does HCP handle machineconfigs anyway? Are they in the management cluster? They are cluster scoped traditionally so how do they separate the different machine configs to each guest cluster?

Reply:
Thanks @JoelSpeed - I've separated `HostedChangeManagementPolicy` (namespaced and for NodePool/HostedCluster) and cluster scoped `ChangeManagementPolicy` (for anything internal to the cluster on which the resource is defined).